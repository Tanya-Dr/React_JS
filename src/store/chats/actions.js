import { SET_CHATS, SET_ERROR } from "./actionTypes";
import { db } from "../../services/firebase";

const setChats = (chats) => ({
  type: SET_CHATS,
  payload: chats,
});

const setError = (error) => ({
  type: SET_ERROR,
  payload: error,
});

export const connectChatsToFB = () => async (dispatch) => {
  try {
    await db.ref("chatList").off();
    await db.ref("chatList").on("value", (snapshot) => {
      let newChats = [];
      snapshot.forEach((snap) => {
        const currentChat = snap.val();
        if (Object.keys(currentChat).length !== 0) {
          newChats.push(currentChat);
        }
      });
      dispatch(setChats(newChats));
    });
  } catch (e) {
    dispatch(setError(e.message));
  }
};

export const addChatWithFB =
  (id, name, robot = false) =>
  async (dispatch) => {
    try {
      await db.ref("chatList").child(id).set({
        id,
        name,
        robot,
      });
    } catch (e) {
      dispatch(setError(e.message));
    }
  };

export const deleteChatWithFB = (id) => async (dispatch) => {
  try {
    await db.ref("chatList").child(id).remove();
  } catch (e) {
    dispatch(setError(e.message));
  }
};
