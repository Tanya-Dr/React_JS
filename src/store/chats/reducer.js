import { ADD_CHAT, DELETE_CHAT } from "./actionTypes";

const initialState = {
  //to be stored like this [{id, name}]
  chatList: [
    { id: "chat1", name: "Chat 1" },
    { id: "chat2", name: "Chat 2" },
    { id: "chat3", name: "Chat 3" },
  ],
};

export const chatsReducer = (state = initialState, { type, payload }) => {
  switch (type) {
    case ADD_CHAT: {
      return {
        ...state,
        chatList: [
          ...state.chatList,
          {
            id: payload.chatId,
            name: payload.name,
          },
        ],
      };
    }
    case DELETE_CHAT: {
      return {
        ...state,
        chatList: [...state.chatList].filter(
          (chat) => chat.id !== payload.chatId
        ),
      };
    }
    default:
      return state;
  }
};
