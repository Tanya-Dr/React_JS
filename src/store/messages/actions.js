import { AUTHORS } from "../../constants";
import {
  ADD_CHAT_TO_MSG_LIST,
  ADD_MESSAGE,
  DELETE_CHAT_FROM_MSG_LIST,
  SET_ERROR_MSG,
  SET_MESSAGES,
} from "./actionTypes";
import { db } from "../../services/firebase";

export const addMessage = (chatId, message) => ({
  type: ADD_MESSAGE,
  payload: {
    chatId,
    message,
  },
});

export const addChatToMsgList = (chatId) => ({
  type: ADD_CHAT_TO_MSG_LIST,
  payload: { chatId },
});

export const deleteChatFromMsgList = (chatId) => ({
  type: DELETE_CHAT_FROM_MSG_LIST,
  payload: { chatId },
});

let timeout;

export const addMessageWithReply = (chatId, message) => (dispatch) => {
  dispatch(addMessage(chatId, message));

  if (timeout) {
    clearTimeout(timeout);
  }

  timeout = setTimeout(() => {
    dispatch(
      addMessage(chatId, {
        id: Date.now(),
        text: "Hello, can I help you?",
        author: AUTHORS.robot,
      })
    );
  }, 1500);
};

const setMessages = (messages) => ({
  type: SET_MESSAGES,
  payload: messages,
});

const setError = (error) => ({
  type: SET_ERROR_MSG,
  payload: error,
});

export const connectMessagesToFB = () => async (dispatch) => {
  try {
    await db.ref("messages").off();
    await db.ref("messages").on("value", (snapshot) => {
      let newMessages = {};
      if (!snapshot) {
        return;
      }
      snapshot.forEach((snap) => {
        const currentMsgs = snap.val();
        if (newMessages[snap.key]) {
          newMessages[snap.key].concat(Object.values(currentMsgs));
        } else {
          newMessages[snap.key] = Object.values(currentMsgs);
        }
      });
      dispatch(setMessages(newMessages));
    });
  } catch (e) {
    dispatch(setError(e.message));
  }
};

export const sendMessageWithFB = (chatId, message) => async (dispatch) => {
  try {
    await db
      .ref("messages")
      .child(chatId)
      .push({
        ...message,
        // id: `${chatId}-${Date.now()}`,
        chatId,
      });
  } catch (e) {
    dispatch(setError(e));
  }
};

export const delMsgFromDelChatWithFB = (chatId) => async (dispatch) => {
  try {
    await db.ref("messages").child(chatId).remove();
  } catch (e) {
    dispatch(setError(e));
  }
};
