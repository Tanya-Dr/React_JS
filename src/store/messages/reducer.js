import { AUTHORS } from "../../constants";
import {
  ADD_CHAT_TO_MSG_LIST,
  ADD_MESSAGE,
  DELETE_CHAT_FROM_MSG_LIST,
  SET_ERROR_MSG,
  SET_MESSAGES,
} from "./actionTypes";

// const initialState = {
//   //to be stored like this {[chatId]:[{id, text, author}]}
//   messageList: {
//     chat1: [{ id: "chat1-1", text: "Hello world!", author: AUTHORS.human }],
//     chat2: [{ id: "chat2-1", text: "This is chat2", author: AUTHORS.human }],
//     chat3: [],
//   },
// };

const initialState = {
  //messageList will be stored like this {[chatId]:[{author, chatId, id, text}]}
  messageList: {},
  error: null,
};

export const messagesReducer = (state = initialState, { type, payload }) => {
  switch (type) {
    // case ADD_MESSAGE: {
    //   const currentList = state.messageList[payload.chatId] || [];
    //   return {
    //     ...state,
    //     messageList: {
    //       ...state.messageList,
    //       [payload.chatId]: [
    //         ...currentList,
    //         {
    //           ...payload.message,
    //         },
    //       ],
    //     },
    //   };
    // }
    // case ADD_CHAT_TO_MSG_LIST: {
    //   return {
    //     ...state,
    //     messageList: {
    //       ...state.messageList,
    //       [payload.chatId]: [],
    //     },
    //   };
    // }
    // case DELETE_CHAT_FROM_MSG_LIST: {
    //   const newState = { ...state };
    //   delete newState.messageList[payload.chatId];
    //   return newState;
    // }
    case SET_MESSAGES: {
      return {
        ...state,
        messageList: payload,
      };
    }
    case SET_ERROR_MSG: {
      return {
        ...state,
        error: payload,
      };
    }
    default:
      return state;
  }
};
