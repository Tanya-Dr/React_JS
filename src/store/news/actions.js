import { REQUEST_ERROR, REQUEST_PENDING, REQUEST_SUCCESS } from "./actionTypes";
import { API_URL } from "../../constants";

export const getNewsPending = () => ({
  type: REQUEST_PENDING,
});

export const getNewsSuccess = (news) => ({
  type: REQUEST_SUCCESS,
  payload: news,
});

export const getNewsError = (error) => ({
  type: REQUEST_ERROR,
  payload: error,
});

export const getNews = () => async (dispatch) => {
  dispatch(getNewsPending());

  try {
    const response = await fetch(API_URL);
    if (!response.ok) {
      throw new Error(`Request failed: ${response.status}`);
    }
    const result = await response.json();

    dispatch(getNewsSuccess(result));
  } catch (err) {
    dispatch(getNewsError(err.message));
  }
};
