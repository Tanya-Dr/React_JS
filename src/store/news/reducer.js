import { REQUEST_STATUS } from "../../constants";
import { REQUEST_ERROR, REQUEST_PENDING, REQUEST_SUCCESS } from "./actionTypes";

const initialState = {
  data: [],
  request: {
    status: REQUEST_STATUS.IDLE,
    error: null,
  },
};

export const newsReducer = (state = initialState, { type, payload }) => {
  switch (type) {
    case REQUEST_PENDING: {
      return {
        ...state,
        data: [],
        request: {
          ...state.request,
          error: null,
          status: REQUEST_STATUS.PENDING,
        },
      };
    }
    case REQUEST_SUCCESS: {
      return {
        ...state,
        data: payload,
        request: {
          ...state.request,
          status: REQUEST_STATUS.SUCCESS,
        },
      };
    }
    case REQUEST_ERROR: {
      return {
        ...state,
        request: {
          ...state.request,
          error: payload,
          status: REQUEST_STATUS.ERROR,
        },
      };
    }
    default:
      return state;
  }
};
