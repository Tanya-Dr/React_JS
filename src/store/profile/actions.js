import { PROFILE_CHANGE_NAME } from "./actionTypes";

export const changeName = (newName) => ({
  type: PROFILE_CHANGE_NAME,
  payload: newName,
});
