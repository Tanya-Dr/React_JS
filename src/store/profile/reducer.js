import {
  PROFILE_CHANGE_NAME,
  PROFILE_TOGGLE_SHOW,
  SET_AUTH,
  SET_ERROR,
} from "./actionTypes";

const initialState = {
  showName: false,
  name: "",
  authorized: false,
  error: null,
};

export const profileReducer = (state = initialState, { type, payload }) => {
  switch (type) {
    case PROFILE_TOGGLE_SHOW: {
      return {
        ...state,
        showName: !state.showName,
      };
    }
    case PROFILE_CHANGE_NAME: {
      return {
        ...state,
        name: payload,
      };
    }
    case SET_AUTH: {
      return {
        ...state,
        authorized: payload,
        error: null,
      };
    }
    case SET_ERROR: {
      return {
        ...state,
        error: payload,
      };
    }
    default:
      return state;
  }
};
