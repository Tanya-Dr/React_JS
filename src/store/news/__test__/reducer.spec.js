import { REQUEST_STATUS } from "../../../constants";
import { getNewsError, getNewsPending, getNewsSuccess } from "../actions";
import { newsReducer } from "../reducer";

describe("newsReducer", () => {
  it("returns state with status loading(1) after getNewsPending action", () => {
    const expected = {
      data: [],
      request: {
        status: REQUEST_STATUS.PENDING,
        error: null,
      },
    };

    const received = newsReducer(undefined, getNewsPending());
    expect(received).toEqual(expected);
  });

  it("returns state with status error(3) and value error after getNewsError action", () => {
    const received = newsReducer(undefined, getNewsError("error"));
    expect(received).toMatchSnapshot();
  });

  it("returns state with status success(2) and list of news in data after getNewsSuccess action", () => {
    const received = newsReducer(undefined, getNewsSuccess(["news"]));
    expect(received).toMatchSnapshot();
  });
});
