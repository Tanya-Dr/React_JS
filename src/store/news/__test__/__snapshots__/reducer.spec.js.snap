// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`newsReducer returns state with status error(3) and value error after getNewsError action 1`] = `
Object {
  "data": Array [],
  "request": Object {
    "error": "error",
    "status": 3,
  },
}
`;

exports[`newsReducer returns state with status success(2) and list of news in data after getNewsSuccess action 1`] = `
Object {
  "data": Array [
    "news",
  ],
  "request": Object {
    "error": null,
    "status": 2,
  },
}
`;
